### PR TITLE
Fix overhaul HTTP retry logic and memory management

### PR DESCRIPTION
**Pull request by ja7ad**
# Pull Request

## What does this PR do?
- optimize HTTP retries, memory safety, and NDJSON parsing

## PR checklist
Please check if your PR fulfills the following requirements:
- [ ] Did you use any AI tool while implementing this PR (code, tests, docs, etc.)? If yes, disclose it in the PR description and describe what it was used for. AI usage is allowed when it is disclosed.
- [ ] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

## Summary by CodeRabbit

* **New Features**
  * Added consolidated client benchmark coverage for GET/POST (small/large), gzip encoding, NDJSON decoding, and retry behavior.
  * Updated benchmark display in the README with new, focused benchmark cases.

* **Bug Fixes**
  * Improved retry reliability by replaying request bodies when possible.
  * Expanded timeout/cancellation classification and optimized NDJSON response decoding.

* **Tests**
  * Reworked request execution tests with broader scenario coverage, plus added edge-case coverage suites.

* **Chores**
  * Minor import-order cleanup in encoding tests.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -322,12 +319,28 @@ func (c *client) sendRequest(
...
-	// Create the HTTP request
+	var bodyBytes []byte
+	if body != nil {
+		bodyBytes, err = io.ReadAll(body)
+		// Close the original body (handles sync.Pool internally if you implement a custom closer, or let GC handle it)
+		_ = body.Close()
+		if err != nil {
+			return nil, fmt.Errorf("unable to read body: %w", err)
+		}
+		body = io.NopCloser(bytes.NewReader(bodyBytes))
```
[the remark this answers is not in the document]
Comment (ja7ad) [reply client.go:330]: While it is true that io.ReadAll will eagerly buffer large files in memory, Meilisearch payloads (usually NDJSON document batches) require this buffering to support automatic retries (502/503 errors are common on cloud platforms). If we disable io.ReadAll for io.Reader sources, any user uploading a batch of documents via an io.Reader will immediately lose retry functionality because the stream cannot be rewound after a failed attempt. The memory overhead is bounded by the Meilisearch server's max payload size (typically 100MB), which is an acceptable temporary heap allocation in Go. A true fix would require a breaking API change to support io.ReadSeeker.

**Review by ElyarSadig — APPROVED** [review]
